### PR TITLE
Python version of the Slack sample

### PR DESCRIPTION
**Pull request by daabr**
Related to - but does not fully resolve - ENG-879 and ENG-899.

Note: TODO in the code (bottom of both program files): ENG-802.

Also adjusted the `.star` file to look more like the `.py` file.

### CODE REVIEW

**Comment by tebeka** [comment]
`munch` looks cool, but I prefer not to add it.

We'll need to support any package that we add to the default venv indefinitely for backward compatibility. That's why I think we should stick to parity with integration and maybe very selected few that are super popular (currently only PyYAML and requests). The user can install whatever they want to the environment if they want to work.

I don't think `munch` passes the threshold of making it to the default enviroment.

@daabr What is the reasoning for adding it to the default environment?

File: slack/program.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    client = slack_sdk.WebClient(token)
+
+    client.auth_test().validate()
+    return client
+
+
+def on_slack_app_mention(event: dict) -> None:
+    """https://api.slack.com/events/app_mention
+
+    Args:
```
Comment (tebeka) [inline slack/program.py:45]: Let's discuss which version of Python documentation do we want to use: Sphinx, Google or numpy (see [here](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html))

Personally I'm used to numpy version, but I don't have a strong perference.

File: slack/program.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    data = _structify(event, "data")
+    user = f"<@{data.user}>"
+    channel = f"<#{data.channel}>"
+    slack = _slack_client()
+
+    # Send messages in response to the event:
+    # - A DM to the user who triggered the event
+    # - Two messages to the channel "#slack-test"
+    # See: https://slack.dev/python-slack-sdk/api-docs/slack_sdk/web/client.html#slack_sdk.web.client.WebClient.chat_postMessage
+    text = f"You mentioned me in {channel} and wrote: `{data.text}`"
```
Comment (tebeka) [inline slack/program.py:57]:
```suggestion
    data = event["data"]
    slack = _slack_client()

    # Send messages in response to the event:
    # - A DM to the user who triggered the event
    # - Two messages to the channel "#slack-test"
    # See: https://slack.dev/python-slack-sdk/api-docs/slack_sdk/web/client.html#slack_sdk.web.client.WebClient.chat_postMessage
    text = f"You mentioned me in <#{channel}> and wrote: `{text}`".format(**data)
```

File: slack/program.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    slack.chat_postMessage(channel=data.user, text=text)
```
Comment (tebeka) [inline slack/program.py:58]: Comment why channel is `data.user`

File: slack/program.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    # For educational purposes, print all the reply objects
+    # in the AutoKitteh session's log.
+    resp.validate()
+    for text in resp.get("messages"):
+        print(text)
+
+
+def on_slack_message(event: dict) -> None:
+    """https://api.slack.com/events/message
```
Comment (tebeka) [inline slack/program.py:105]: This URL doesn't explain what the handler does.

File: slack/program.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    text %= (user, data.parent_user_id, data.text)
+    _slack_client().chat_postMessage(
+        channel=data.channel, text=text, thread_ts=data.thread_ts
+    )
+
+
+def _on_slack_message_changed(data: munch.Munch, user: str) -> None:
+    """Someone edited a message."""
+    text = ":point_up: %s edited a message from `%s` to `%s`"
+    text %= (user, data.previous_message.text, data.message.text)
```
Comment (tebeka) [inline slack/program.py:141]: This uses old style formatting, keep it consistent and use new style or f-strings.

File: slack/program.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    """
+    data = _structify(event, "data")
+    slack = _slack_client()
+
+    # Retrieve the profile information of the user who triggered this event.
+    # See: https://slack.dev/python-slack-sdk/api-docs/slack_sdk/web/client.html#slack_sdk.web.client.WebClient.users_info
+    user_info = slack.users_info(user=data.user_id)
+
+    # Encountered an error? Print debugging information
+    # in the AutoKitteh session's log, and finish.
```
Comment (tebeka) [inline slack/program.py:176]: Combine to a single line

**Review by tebeka — CHANGES_REQUESTED** [review]

File: slack/README.md [hunk 7]
```diff
@@ -1,38 +1,54 @@
...
+The code files ([`program.py`](./program.py) or [`program.star`](./program.star))
+implement multiple entry-point functions that are triggered by incoming Slack
+events, which are defined in the [`autokitteh-python.yaml`](./autokitteh-python.yaml)
+or [`autokitteh-starlark.yaml`](./autokitteh-starlark.yaml) manifest files.
+These functions also execute various Slack API calls.
 
 API details:
 
 - [Web API reference](https://api.slack.com/methods)
+- [Python web API reference](https://slack.dev/python-slack-sdk/api-docs/slack_sdk/web/client.html)
```
[the remark this answers is not in the document]
Comment (daabr) [reply slack/README.md:18]: done

File: slack/program.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+"""
+
+import os
+import time
+
+import munch
+import slack_sdk
+
+
+def _structify(event: dict, key: str = "") -> munch.Munch:
```
[the remark this answers is not in the document]
Comment (daabr) [reply slack/program.py:23]: 1. Indeed, let's discuss the typing topic offline
2. My rule of thumb is that if a function is shorter than the would-be docstring, and is trivial to understand, the docstring isn't needed; this is supported by Black's code checks which don't require a docstring for less than 5 LOC

File: slack/program.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+- Events API reference: https://api.slack.com/events?filter=Events
+
+This program isn't meant to cover all available functions and events.
+It merely showcases various illustrative, annotated, reusable examples.
+"""
+
+import os
+import time
+
+import munch
```
[the remark this answers is not in the document]
Comment (daabr) [reply slack/program.py:19]: I agree in general - I want Python samples to focus on the thing that they're demonstrating, i.e. integrations in existing samples (except PuRRR and future functional-goal-oriented samples).

However, event data is provided as a dictionary, and I don't consider that readable, so I'm forced to work around that with Munch. See my reply to your suggested change below.

For example, consider the unreadability of:

```python
slack.chat_update(channel=resp["channel"], ts=resp["ts"], text="Blah blah blah `{data[\"previous_message\"][\"text\"]}` blah `{data[\"message\"][\"text\"]}`")
```

...And rememeber that this is a **simple** example. Real-life usage is expected to be even more complex.

File: slack/program.py (hunk 2 above)
Comment (daabr) [reply slack/program.py]: Resolved by https://github.com/autokitteh/autokitteh/pull/341

File: slack/program.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (daabr) [reply slack/program.py]: Even though I like type hinting a lot, I'm removing it for now. We'll rethink it if someone asks for it.

File: slack/program.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tebeka) [reply slack/program.py]: > I still need to modify responses from the Slack SDK, but it's much simpler.

You mean one from the Python SDK? This is the experience Python users will have with the Slack SDK. Up to them what to do with it. I prefer to keep it as close to the SDK as possible.

File: slack/program.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+import slack_sdk
+
+
+def _structify(event: dict, key: str = "") -> munch.Munch:
+    if key:
+        event = event[key]
+    return munch.Munch.fromDict(event)
+
+
+def _slack_client() -> slack_sdk.WebClient:
```
[the remark this answers is not in the document]
Comment (tebeka) [reply slack/program.py:23]: I prefer to call handler `on_event` or something like that, but go ahead and use `_`.

File: slack/program.py (hunk 4 above)
Comment (tebeka) [reply slack/program.py:95]: I meant comment like: `"event" comes from slack SDK, see https://api.slack.com/events/message for more information`

File: slack/program.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (daabr) [reply slack/program.py]: Agreed, but I'll do it in a separate PR.

**Pull request merged**